datalad/support/protocol.py

```diff
@@ -0,0 +1,319 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+""" Protocolling  command calls.
+"""
+
+from abc import ABCMeta, abstractmethod, abstractproperty
+from os import linesep
+import logging
+import time
+
+lgr = logging.getLogger('datalad.protocol')
+
+
+class ProtocolInterface(object):
+    """Interface class for protocols used by the Runner.
+
+    Implementations of this interface are supposed to store one section per
+    recorded command call in `self._sections` as a dictionary. Default
+    implementation of __str__ has to be overridden otherwise.
+
+    ProtocolInterface is iterable as a list of sections.
+    """
+
+    __metaclass__ = ABCMeta
```

Comment:
Yeah - I like to have an option to enforce implementation of a certain interface finally.